language-plutus-core/src/Language/PlutusCore/Instance/Eq/Internal.hs

```diff
@@ -0,0 +1,263 @@
+
+instance Semigroup a => Semigroup (Bilateral a) where
+    Bilateral l1 r1 <> Bilateral l2 r2 = Bilateral (l1 <> l2) (r1 <> r2)
+
+instance Monoid a => Monoid (Bilateral a) where
+    mempty = Bilateral mempty mempty
+
+-- To rename from left to right is to update the left renaming.
+instance HasRenaming ren unique => HasRenaming (Bilateral ren) (LR unique) where
+    renaming = bilateralL . renaming . coerced @(Renaming unique)
+
+-- To rename from right to left is to update the right renaming.
+instance HasRenaming ren unique => HasRenaming (Bilateral ren) (RL unique) where
+    renaming = bilateralR . renaming . coerced @(Renaming unique)
+
+-- | The type of a runnable equality check. @Maybe ()@ is isomorphic to 'Bool' and we use it
+-- instead of 'Bool', because this unlocks the convenient and readable do-notation and allows for
+-- automatic short-circuiting, which would be tedious with @Rename (Bilateral ren) Bool@.
+type EqRename ren = RenameT (Bilateral ren) Maybe ()
+type ScopedEqRename = EqRename ScopedRenaming
+
+-- | Run an 'EqRename' computation.
+runEqRename :: Monoid ren => EqRename ren -> Bool
+runEqRename = isJust . runRenameT
+
+{- Note [Modulo alpha]
+The implemented algorithm of checking equality modulo alpha works as follows
+(taking types as an example):
+
+- traverse both the types simultaneously
+- if the outermost constructors differ, return 'False'
+- otherwise if the constructors are binders, then record that the two possibly distinct bound
+    variables map to each other
+- otherwise if the constructors are variables, look them up in the current scope
+    * if both are in scope, then those are bound variables, so check that they map to each other
+        (i.e. are introduced by twin binders)
+    * if both are not in scope, then those are free variables, so check their direct equality
+    * if one is in scope and the other one is not, then return 'False'
+- otherwise check equality of non-recursive constituents and recurse for recursive occurrences
+-}
+
+{- Note [Scope tracking]
+Scopes (term level vs type level) are resolved automatically in a type-driven way much like we do
+in the renamer. This allows not to think about scopes when implementing the actual equality checks
+and makes it impossible to confuse the scopes and e.g. insert a type-level name into a term-level
+scope. Also allows to define a function that records bindings and a function that checks equality
+of two names in a general manner for both the scopes.
+-}
+
+{- Note [Side tracking]
+The simplest way to track that twin variables map to each other is to have two contexts:
+
+- left-hand side variables and what they map to
+- right-hand side variables and what they map to
+
+E.g. when checking equality of these two types:
+
+    all (x_3 :: *) (x_3 :: *). x_3 -> x_3
+    all (y_4 :: *) (z_5 :: *). y_4 -> z_5
+
+we first record that @x_3@ maps to @y_4@ and vice versa, then record that @x_3@ maps to @z_5@ and
```

Comment:
Where do these mappings come from (RHS or LHS)? I assume they get stored in both RL and LR, but that'd be nice to add.

Reply:
Updated. Please check whether the note reads better now.

Reply:
Doesn't look too different to me?

Reply:
I added these sentences:

> For storing the left-to-right and right-to-left mappings separately we use the 'Bilateral' data
type.

> I.e. you wrap a name into either 'LR' and 'RL' and depending on that you get focused on either left-to-right or right-to-left part of a 'Bilateral' renaming.

So `LR` and `RL` do not store any renamings. Renamings are stored in `Bilateral`. `LR` and `RL` just let the user be explicit about whether they want to update from left to right or from right to left.

Reply:
I think my main confusion is about not having a proper definition of "left-to-right" and "right-to-left". Are they related to left-hand-side / right-hand-side of an expression?

Reply:
We are comparing two terms, and matching up variables between them: there is therefore a bijective mapping between names that appear in the one term ("the left") and names that appear in the other term ("the right"). Left-to-right and right-to-left refer to the two directions of the mapping.

Reply:
Added the following comment:

> (we refer to a first argument as being on the left-hand side and to a second argument as being on the right-hand side)